### ISSUE TITLE
question surrounding desired `scipy.special._legacy` deprecations

### ISSUE THREAD

**Issue by grlee77** [scipy/scipy#15596]
When working on porting some `scipy.special` functions to CuPy, I noticed that there was a scheduled deprecation for SciPy 1.7 that has not yet been carried out. Specifically, the unsafe double->int casting for the `n` or `k` inputs to binomial distribution functions `bdtr`, `bdtrc` and `bdtri` leads to the following warning:

https://github.com/scipy/scipy/blob/b5d8bab88af61d61de09641243848df63380a67f/scipy/special/_legacy.pxd#L39-L43

I thought about making a PR to complete the deprecation, however, I am not sure how best to proceed with that as it seems inconsistent with other functions in that same file. For example the functions related to negative binomial distributions (`nbdtr`, `nbdtrc` and `nbdtri`) and others instead raise a `RuntimeWarning` and do not indicate that the unsafe casting is deprecated. Shouldn't the approach for `bdtr` and `nbdtr` be the consistent? Please advise.

**Comment by rgommers** [comment]
Most functions now do `_legacy_cast_check` rather than `_legacy_deprecation`. Do I understand correctly that is we moved them all to `_legacy_deprecation`, then things are consistent? That seems like a reasonable thing to do.

**Comment by grlee77** [comment]
That would be consistent,  however `_legacy_deprecation` was introduced in `v1.5` and still mentions removal in `v1.7.x` specifically. If we change everything to `_legacy_deprecation` then would we list it as `v1.10`?

**Comment by rgommers** [comment]
We'd then be deprecating everything in `1.9.0` so should remove it in `1.11.0`.

**Comment by grlee77** [comment]
For historical context, https://github.com/scipy/scipy/pull/11045 is the PR included in SciPy 1.5 where the safe ufunc type signatures for these `bdtr` functions were changed from `iid->d` to `did->d` and the `_legacy_deprecation` was introduced.

Specifically, reasoning for wanting to allow double-valued `k` from a CDF perspective: https://github.com/scipy/scipy/pull/11045#issuecomment-553068782

However, these binomial distribution funtions are now inconsistent with the corresponding negative binomial ones (`nbdtr`, `nbdtrc`, `nbdtri`) whose signature [was not also changed](https://github.com/scipy/scipy/blob/bcfc86dcc25fbf26155fd5fa2cf2451a4a405e2b/scipy/special/functions.json#L1040) to `did->d` in the same manner.

**Comment by h-vetinari** [comment]
I unintentionally duplicated this issue in https://github.com/scipy/scipy/issues/15731 - my opinion was that since the warning explicitly [announces](https://github.com/scipy/scipy/blob/v1.8.0/scipy/special/_legacy.pxd#L41-L43) removal by 1.7, we could directly do this in 1.9.

**Comment by grlee77** [comment]
> I unintentionally duplicated this issue in https://github.com/scipy/scipy/issues/15731 - my opinion was that since the warning explicitly [announces](https://github.com/scipy/scipy/blob/v1.8.0/scipy/special/_legacy.pxd#L41-L43) removal by 1.7, we could directly do this in 1.9.

My only problem with that is that it would be nice if the negative binomial functions (`nbdtr*`) would deprecate in the same way as the binomial ones (`bdtr*`). Given that we don't currently have a similar deprecation warning for the `nbdtr*` functions, we could modify those to behave like `bdtr*` now and then just deprecate both together in 1.11.

**Comment by h-vetinari** [comment]
Yeah, we're not in a rush. Fine by me to do it for 1.11. Can you change the title of this issue to reflect the plan to deprecate and what? Then I'll close #15731 and update the tracker.

**Comment by h-vetinari** [comment]
Gentle ping @grlee77 for renaming this issue... 🙃

**Comment by grlee77** [comment]
> Gentle ping @grlee77 for renaming this issue... upside_down_face

done, but feel free to modify it if you like. I can try to submit a small PR to address this later in the week.

**Comment by h-vetinari** [comment]
Thanks; I don't have the rights to rename issues. 🙃

**Comment by h-vetinari** [comment]
Removing the "needs-decision" label, as a way forward seems to have been agreed upon. Please revert if I'm mistaken.

**Comment by h-vetinari** [comment]
@grlee77, any update on this?

**Comment by h-vetinari** [comment]
Gentle ping @grlee77

**Comment by h-vetinari** [comment]
Update: this should not be deprecated but switched to the boost implementation. For more details see #15845

**Comment by tylerjereddy** [comment]
I'm guessing the Boost implementation won't be ready in 24 hours, bumping milestone for imminent branching on May 29th.

**Comment by h-vetinari** [comment]
CC @steppi @izaid, not sure if this is on your radar, but looks like it'd be closely related to recent C++ efforts in scipy.special.

**Comment by steppi** [comment]
> CC @steppi @izaid, not sure if this is on your radar, but looks like it'd be closely related to recent C++ efforts in scipy.special.

Reading through the above, things seem to have come full circle. @izaid and I are actually considering to move away from using Boost in `special` because it is more difficult to make the Boost implementations work on CUDA (in CuPy for instance). To fill anyone in who might not be aware; @izaid and I are trying to assemble the scalar kernels in `scipy.special` into a header only C++ library which can run on CPU or on GPU with CUDA. This was first discussed here, https://github.com/scipy/scipy/issues/19404, and here's the first example of this in action: https://github.com/scipy/scipy/pull/19601/files, https://github.com/cupy/cupy/pull/8140. This would allow SciPy, CuPy, and other array libraries like PyTorch to share code for their special function implementations.

I've recently translated `nbdtr` and friends to C++ in this PR, https://github.com/scipy/scipy/pull/20390. Rather than using Boost, @izaid and I hope to update the existing implementations of these functions to reach parity with Boost.

**Comment by h-vetinari** [comment]
The title here is somewhat fungible 😅; it's already gone through
* question surrounding desired `scipy.special._legacy deprecation`
* reschedule `scipy.special.bdtr` deprecations and make `scipy.special.nbdtr` follow the same behavior
* ENH: special: update nbdtr and friends to use Boost implementation

So I'm pretty sure we can close this once #20390 lands. :)

See also this [comment](https://github.com/scipy/scipy/pull/15845#issuecomment-1563297333) from #15845:
> @tylerjereddy: I spoke with Matt [Haberland] in person this morning and he had a pretty firm view that we should not do this deprecation and instead pursue the Boost approach.
> 
> I spoke with Greg [Lee] yesterday afternoon and I think this PR had fallen off his radar for ~1 year and he already seemed "OK' with going the Boost route above in https://github.com/scipy/scipy/pull/15845#issuecomment-1134579184.

**Comment by lucascolley** [comment]
@steppi how are we looking with the new implementations with respect to Matt's comment from the linked PR?

> I see that currently `nbdtr` and friends truncate floating point `n` to an integer, and that's not very useful. But instead of deprecating, have we considered using Boost (see e.g. gh-14696) to perform the calculation instead? It (meaningfully) extends the negative binomial distribution to float `n`, so that's what we do in `stats`:
```python3
from scipy.stats import nbinom
nbinom.cdf(12, p = 0.5, n=15)  # 0.3505540192127228
nbinom.cdf(12, p = 0.5, n=15.5)  # 0.31720141151535164
```

> (Agrees with Wolfram Alpha [CDF[NegativeBinomialDistribution[15.5, 0.5], 12]](https://www.wolframalpha.com/input?i=CDF%5BNegativeBinomialDistribution%5B15.5%2C+0.5%5D%2C+12%5D) and R [`pnbinom(12, 15.5, 0.5)`](https://rdrr.io/snippets/))

If there is still work to do there, we should probably change the title to represent what is left to do, and remove this from the deprecation tracker.

**Comment by lucascolley** [comment]
Okay, this is not resolved yet:

```python
In [1]: from scipy import special, stats

In [2]: stats.nbinom.cdf(12, p = 0.5, n=15)
Out[2]: np.float64(0.3505540192127228)

In [3]: special.nbdtr(12, 15, 0.5)
Out[3]: np.float64(0.35055401921272267)

In [4]: stats.nbinom.cdf(12, p = 0.5, n=15.5)
Out[4]: np.float64(0.31720141151535153)

In [5]: special.nbdtr(12, 15.5, 0.5)
<ipython-input-5-aa33b818cad8>:1: RuntimeWarning: floating point number truncated to an integer
  special.nbdtr(12, 15.5, 0.5)
Out[5]: np.float64(0.35055401921272267)
```

**Comment by dschmitz89** [comment]
Since Boost started supporting CUDA recently, I will migrate the binomial and negative binomial special function families to boost in the coming weeks.

**Comment by rgommers** [comment]
> Since Boost started supporting CUDA recently,

That should matter to CuPy et al., but not to SciPy? Either way, migrating to Boost if that has a better implementation SGTM.

**Comment by lucascolley** [comment]
any update here @dschmitz89 ?

**Comment by dschmitz89** [comment]
Not gonna make it in this release. Could not get the boost wrapper working with integers within a few minutes when I tried a few weeks back unfortunately.
